python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,315 @@
+import gc
+import heapq
+import logging
+import numbers
+import os
+import shutil
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Dict, Union, Callable, Tuple, List, Any
+
+import ray
+from ray.tune.result import NODE_IP, TRAINING_ITERATION
+from ray.util import PublicAPI
+from ray.util.ml_utils.util import is_nan
+
+MAX = "max"
+MIN = "min"
+
+logger = logging.getLogger(__name__)
+
+
+class _TrackedCheckpoint:
+    MEMORY = "memory"
+    PERSISTENT = "persistent"
+
+    def __init__(
+        self,
+        checkpoint_dir_or_data: Optional[Union[str, Path, Dict, ray.ObjectRef]],
+        storage_mode: str,
+        checkpoint_id: Optional[int] = None,
+        result: Optional[Dict] = None,
+        node_ip: Optional[str] = None,
+    ):
+        self.checkpoint_dir_or_data = checkpoint_dir_or_data
+        self.checkpoint_id = checkpoint_id
+        self.storage_mode = storage_mode
+
+        # Todo: What to do if result is a subset of checkpoint_dir_or_data (dict)
+        self.result = result or {}
+        self.node_ip = node_ip or self.result.get(NODE_IP, None)
+
+    def commit(self, path: Optional[Path] = None):
+        """Commit checkpoint to disk, if needed."""
+        pass
+
+    def delete(
+        self, delete_fn: Optional[Callable[["_TrackedCheckpoint"], None]] = None
+    ):
+        """Delete checkpoint from disk, if needed."""
+        delete_fn = delete_fn or _default_delete_fn
+        delete_fn(self)
+
+    def __repr__(self):
+        if self.storage_mode == _TrackedCheckpoint.MEMORY:
+            return f"<_TrackedCheckpoint storage='MEMORY' result={self.result}>"
+
+        return (
+            f"<_TrackedCheckpoint storage='PERSISTENT' "
+            f"checkpoint_dir_or_data={self.checkpoint_dir_or_data}>"
+        )
+
+
+def _default_delete_fn(checkpoint: _TrackedCheckpoint):
+    if checkpoint.storage_mode != _TrackedCheckpoint.PERSISTENT:
+        return
+
+    if isinstance(checkpoint.checkpoint_dir_or_data, (str, bytes, os.PathLike)):
+        if os.path.isfile(checkpoint.checkpoint_dir_or_data):
+            os.remove(checkpoint.checkpoint_dir_or_data)
+            return
+        elif os.path.isdir(checkpoint.checkpoint_dir_or_data):
+            shutil.rmtree(checkpoint.checkpoint_dir_or_data)
+            return
+    logger.warning(
+        f"Could not delete checkpoint {checkpoint} from disk as it is "
+        f"neither file not directory. Path: {checkpoint.checkpoint_dir_or_data}."
+    )
```

Comment:
add error handling/ignore errors?

Reply:
Also to other places where we remove files/folders

Reply:
I wrapped the whole delete_fn call in an except block now which I think should be sufficient.